agent/lib/kontena/watchdog.rb

```diff
@@ -0,0 +1,112 @@
+# Run the given block in a separate Thread every interval, and abort if it times out or fails
+class Kontena::Watchdog
+  include Celluloid
+  include Kontena::Logging
+
+  INTERVAL = 10.0
+  THRESHOLD = 10.0
+  TIMEOUT = 30.0
+
+  # this is not a StandardError, it is supposed to abort the thread
+  class Abort < Exception
+
+  end
+
+  def self.watch(**options, &block)
+    new(Thread.current, block, **options)
+  end
+
+  def initialize(thread, block, interval: INTERVAL, threshold: THRESHOLD, timeout: TIMEOUT, abort: true)
+    @thread = thread
+    @block = block
+
+    @interval = interval
+    @threshold = threshold
+    @timeout = timeout
+    @abort = abort
+
+    async.start
+  end
+
+  def start
+    info "watchdog start"
+    @timer = every(@interval) do
+      if !@ping || (@pong && @ping < @pong)
+        async.ping
+      elsif !@pong || @pong < @ping
+        check
+      else
+
+      end
+    end
+  end
+
+  # Stop the every loop from start
+  def stop
+    @timer.cancel
+  end
+
+  # Start new watchdog ping
+  def ping
+    ping = @ping = Time.now
+
+    defer {
+      @ping_thread = Thread.current
+
+      @block.call
+
+      @ping_thread = nil
+    }
+
+  rescue => exc
+    abort(exc)
+  else
+    pong(ping, Time.now)
+  end
+
+  def pong(ping, pong)
+    @pong = pong if @ping == ping
+  end
+
+  # Watchdog has not yet seen any @pong for the latest @ping, warn on threshold and abort on timeout
+  def check
+    delay = Time.now - @ping
+
+    if delay > @timeout
+      bite(delay)
+    elsif delay > @threshold
+      bark(delay)
+    else
+      debug { "watchdog delay is %.3fs" % [delay] }
+    end
+  end
+
+  # @return [Array<String>] current target thread stack
+  def trace
+    @ping_thread ? @ping_thread.backtrace : []
+  end
+
+  # warn when @ping delay exceeds @threshold
+  # @param delay [Float] > @threshold
+  def bark(delay)
+    warn "watchdog delayed by %.3fs @ %s" % [delay, trace.join("\n\t")]
+  end
+
+  # abort when @ping delay exceeds @timeout
+  # @param delay [Float] > @timeout
+  def bite(delay)
+    exc = Timeout::Error.new "watchdog timeout after %.3fs @ %s" % [delay, trace.join("\n\t")]
+    abort(exc)
+  end
+
+  # kill the watched Thread
+  def abort(exc)
+    error exc
+
+    # only abort once
+    stop
+
+    # assume that the thread has abort_on_exception and it does not rescue non-StandardError
+    @thread.raise Abort, "#{exc.class}: #{exc.message}" if @abort
```

Comment:
May want to abort harder... Celluloid traps the `Abort` in the main thread and does its own shutdown, which is not really what we want in the case of watchdog failures:

```
E, [2017-12-13T15:20:58.464693 #1] ERROR -- Kontena::Watchdog: watchdog timeout after 39.094s @ /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `sleep'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `block in check'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:14:in `for'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:58:in `check'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:77:in `block in receive'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:14:in `for'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:76:in `receive'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:50:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid.rb:141:in `suspend'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:41:in `response'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:45:in `value'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/proxy/sync.rb:22:in `method_missing'
	/app/lib/kontena/agent.rb:145:in `block in watchdog'
	/app/lib/kontena/watchdog.rb:56:in `block in ping'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/future.rb:18:in `block in new'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-essentials-0.20.5/lib/celluloid/internals/thread_handle.rb:14:in `block in initialize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor/system.rb:78:in `block in get_thread'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:50:in `block in instantiate' (Timeout::Error)

I, [2017-12-13T15:20:58.612649 #1]  INFO -- Kontena::Workers::LogWorker: stop log streaming
I, [2017-12-13T15:20:58.624217 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: stopped to check status from /ingress-lb.lb-1
E, [2017-12-13T15:21:07.721385 #1] ERROR -- : Couldn't cleanly terminate all actors in 10 seconds!
/app/lib/kontena/agent.rb:98:in `gets': Timeout::Error: watchdog timeout after 39.094s @ /usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `sleep' (Kontena::Watchdog::Abort)
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:63:in `block in check'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:14:in `for'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:58:in `check'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:77:in `block in receive'
	/usr/lib/ruby/gems/2.4.0/gems/timers-4.1.1/lib/timers/wait.rb:14:in `for'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/mailbox.rb:76:in `receive'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:50:in `wait'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid.rb:141:in `suspend'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:41:in `response'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:45:in `value'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/proxy/sync.rb:22:in `method_missing'
	/app/lib/kontena/agent.rb:145:in `block in watchdog'
	/app/lib/kontena/watchdog.rb:56:in `block in ping'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/future.rb:18:in `block in new'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-essentials-0.20.5/lib/celluloid/internals/thread_handle.rb:14:in `block in initialize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor/system.rb:78:in `block in get_thread'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:50:in `block in instantiate'
	from /app/lib/kontena/agent.rb:98:in `run!'
	from /app/bin/kontena-agent:46:in `<main>'
```